app/src/main/java/org/thoughtcrime/securesms/jobmanager/JobManager.java

```diff
@@ -0,0 +1,635 @@
+   *
+   * @return If the job completed, this will contain its completion state. If it timed out or
+   *         otherwise didn't complete, this will be absent.
+   */
+  @WorkerThread
+  public Optional<JobTracker.JobState> runSynchronously(@NonNull Job job, long timeout) {
+    CountDownLatch                       latch       = new CountDownLatch(1);
+    AtomicReference<JobTracker.JobState> resultState = new AtomicReference<>();
+
+    addListener(job.getId(), new JobTracker.JobListener() {
+      @Override
+      public void onStateChanged(@NonNull Job job, @NonNull JobTracker.JobState jobState) {
+        if (jobState.isComplete()) {
+          removeListener(this);
+          resultState.set(jobState);
+          latch.countDown();
+        }
+      }
+    });
+
+    add(job);
+
+    try {
+      if (!latch.await(timeout, TimeUnit.MILLISECONDS)) {
+        return Optional.absent();
+      }
+    } catch (InterruptedException e) {
+      Log.w(TAG, "Interrupted during runSynchronously()", e);
+      return Optional.absent();
+    }
+
+    return Optional.fromNullable(resultState.get());
+  }
+
+  /**
+   * Retrieves a string representing the state of the job queue. Intended for debugging.
+   */
+  @WorkerThread
+  public @NonNull String getDebugInfo() {
+    AtomicReference<String> result = new AtomicReference<>();
+    CountDownLatch          latch  = new CountDownLatch(1);
+
+    runOnExecutor(() -> {
+      result.set(jobController.getDebugInfo());
+      latch.countDown();
+    });
+
+    try {
+      boolean finished = latch.await(10, TimeUnit.SECONDS);
+      if (finished) {
+        return result.get();
+      } else {
+        return "Timed out waiting for Job info.";
+      }
+    } catch (InterruptedException e) {
+      Log.w(TAG, "Failed to retrieve Job info.", e);
+      return "Failed to retrieve Job info.";
+    }
+  }
+
+  /**
+   * Adds a listener that will be notified when the job queue has been drained.
+   */
+  void addOnEmptyQueueListener(@NonNull EmptyQueueListener listener) {
+    runOnExecutor(() -> {
+      synchronized (emptyQueueListeners) {
+        emptyQueueListeners.add(listener);
+      }
+    });
+  }
+
+  /**
+   * Removes a listener that was added via {@link #addOnEmptyQueueListener(EmptyQueueListener)}.
+   */
+  void removeOnEmptyQueueListener(@NonNull EmptyQueueListener listener) {
+    runOnExecutor(() -> {
+      synchronized (emptyQueueListeners) {
+        emptyQueueListeners.remove(listener);
+      }
+    });
+  }
+
+  @Override
+  public void onConstraintMet(@NonNull String reason) {
+    Log.i(TAG, "onConstraintMet(" + reason + ")");
+    wakeUp();
+  }
+
+  /**
+   * Blocks until all pending operations are finished.
+   */
+  @WorkerThread
+  public void flush() {
+    CountDownLatch latch = new CountDownLatch(1);
+
+    runOnExecutor(latch::countDown);
+
+    try {
+      latch.await();
+      Log.i(TAG, "Successfully flushed.");
+    } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)